docs/js-api-reference.rst

```diff
@@ -25,5 +25,33 @@ You can overwrite any of them, if any kind of customizations are required.
 .. js:autofunction:: showSearchModal
 .. js:autofunction:: removeSearchModal
 
+Example: Overwriting A JavaScript Function
+------------------------------------------
+
+Here is an example on how to add some customizations
+by overriding some of the above functions.
+
+Let's say you have a custom search bar (different from the default one),
+and you want to show the search UI when the user clicks on that.
+For this, you can override ``getInputField()`` function to return your own search bar.
+This will add all the required event-listeners in the right place.
+
+Create a ``custom.js`` and write the following code:
+
+.. code-block:: js
+
+    // custom.js
+
+    /**
+     * Return the required search bar.
+     */
+    function getInputField() {
+        var custom_search_bar = document.querySelector('.my-custom-search-bar');
+        return custom_search_bar;
+    }
+
+And then, you can `include this custom.js file in your docs`_.
```

Comment:
Does it have to be included after our file? I don't quite understand how this is working -- are we not namespacing our functions at all?

Reply:
I tested this approach and there was this problem:
- The custom js file is included before rtd_sphinx_search.js

Then to overwrite a function, I found a way:
In our `rtd_sphinx_search.js`, wrap every function like this:

```js
if (!window.hasOwnProperty("function_name")) {
    window.function_name = () => {
        // function definition
    }
```

This approach works, I tested this locally.

Reply:
I haven't done namespacing...
Should I add `RTD` before every variable?

Reply:
Almost certainly we should be namespacing this code. I don't know the best way to do this in JS, @davidfischer? I think probably with some kind of object that contains all the functions.

Reply:
We definitely want to namespace. I would go with `READTHEDOCS` so functions would be `READTHEDOCS. getInputField()`.

You can check whether `READTHEDOCS` exists with `typeof READTHEDOCS === 'undefined'`. Generally the order of JavaScript being included can be controlled but even then you can always wait until DOM ready to actually override anything.

The major reason for this is to not pollute the global namespace and to limit all our methods and data to one place. I would also like to eventually put `READTHEDOCS_DATA` into there (eg. `READTHEDOCS.BUILD_DATA`).